src/core/processing/qgsprocessingutils.cpp

```diff
@@ -172,3 +173,29 @@ QString QgsProcessingUtils::normalizeLayerSource( const QString &source )
   return normalized.trimmed();
 }
 
+QgsFeatureIterator QgsProcessingUtils::getFeatures( QgsVectorLayer *layer, const QgsProcessingContext &context, const QgsFeatureRequest &request )
+{
+  bool useSelection = context.flags() & QgsProcessingContext::UseSelection && layer->selectedFeatureCount() > 0;
+
+  QgsFeatureRequest req( request );
+  req.setInvalidGeometryCheck( context.invalidGeometryCheck() );
+  req.setInvalidGeometryCallback( context.invalidGeometryCallback() );
+  if ( useSelection )
+  {
+    return layer->selectedFeaturesIterator( req );
+  }
+  else
+  {
+    return layer->getFeatures( req );
+  }
+}
+
+long QgsProcessingUtils::featureCount( QgsVectorLayer *layer, const QgsProcessingContext &context )
+{
+  bool useSelection = context.flags() & QgsProcessingContext::UseSelection && layer->selectedFeatureCount() > 0;
```

Comment:
The `layer->selectedFeatureCount() > 0` condition might lead to surprising results in algorithms that set a selection and then continue to process based on this selection. If a select by rectangle doesn't select anything, this will result in processing all features instead of none.